### PR TITLE
MGMT-16090: allow PostgreSQL upgrades

### PR DESCRIPTION
**Pull request by eifrach**
## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 1]
```diff
@@ -1634,6 +1634,10 @@ func newAssistedServiceDeployment(ctx context.Context, log logrus.FieldLogger, a
 			newSecretEnvVar("POSTGRESQL_DATABASE", "db.name", databaseName),
 			newSecretEnvVar("POSTGRESQL_USER", "db.user", databaseName),
 			newSecretEnvVar("POSTGRESQL_PASSWORD", "db.password", databaseName),
+			// allow auto upgrade when new image is available
+			// allows upgrade for +1 version only, and cannot skip versions
+			newSecretEnvVar("POSTGRESQL_UPGRADE", "POSTGRESQL_UPGRADE", "true"),
```
Comment (rccrdpccl) [inline internal/controller/controllers/agentserviceconfig_controller.go:1639]: most likely not from secret

Comment (carbonin) [reply internal/controller/controllers/agentserviceconfig_controller.go:1639]: Right, you should just define the `corev1.EnvVar` directly here.

**Review by carbonin — COMMENTED** [review]
Can we also update the image we use as our default and for tests in this PR?

**Comment by eifrach** [comment]
> the image we use as our default and for

yes, I will change the to the [sql-13 ](https://github.com/openshift/assisted-service/blob/184bc015126388786312c97e5daf87180f56c1af/internal/controller/controllers/images.go#L36) befoe stating the tests.
I want to deploy it as is and redeploy it with the new image to see the migration works

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 2]
```diff
@@ -1661,6 +1661,16 @@ func newAssistedServiceDeployment(ctx context.Context, log logrus.FieldLogger, a
 			newSecretEnvVar("POSTGRESQL_DATABASE", "db.name", databaseName),
 			newSecretEnvVar("POSTGRESQL_USER", "db.user", databaseName),
 			newSecretEnvVar("POSTGRESQL_PASSWORD", "db.password", databaseName),
+			// allow auto upgrade when new image is available
+			// allows upgrade for +1 version only, and cannot skip versions
+			corev1.EnvVar{
+				Name:  "POSTGRESQL_UPGRADE",
```
Comment (rccrdpccl) [inline internal/controller/controllers/agentserviceconfig_controller.go:1667]: What other options we have other than hardlink? Would it change depending on what image we run? Could it be worth having it as env var with default value? So people can technically tweak it (maybe in the next releases of MCE if it's too tight for this one)

Comment (eifrach) [reply internal/controller/controllers/agentserviceconfig_controller.go:1667]: you have two options
copy: which need 2 PVC and it copies data from PV1 to PV2 as upgraded version
hardlink: migrate the data to the upgrade version 

we agreed on using the `hardlink` way

File: deploy/postgres/postgres-deployment-ephemeral.yaml [hunk 3]
```diff
@@ -1,3 +1,27 @@
...
+    fi
+
+    BINARY_VERSION=$(postgres --version | grep -Eo '[0-9].')
+    DB_VERSION=$(cat $PGDATA/PG_VERSION)
+
+    if [[ $BINARY_VERSION > $DB_VERSION ]]; then
+      # clear PID for the upgrade
+      echo "$0: found postgres version $BINARY_VERSION newer than database version $DB_VERSION"
+      echo "$0: Removing postmaster pid file"
+      rm -f cat /var/lib/pgsql/data/userdata/postmaster.pid
```
[the remark this answers is not in the document]
Comment (eifrach) [reply deploy/postgres/postgres-deployment-ephemeral.yaml:22]: so even when you do it manually and shutdown gracefully -- the postmaster is not deleting

**Pull request closed without merging**